src/daft-scan/src/scan_task_iters.rs

```diff
@@ -28,35 +28,80 @@ type BoxScanTaskIter<'a> = Box<dyn Iterator<Item = DaftResult<ScanTaskRef>> + 'a
 /// * `max_size_bytes`: Maximum size in bytes of a ScanTask, capping the maximum size of a merged ScanTask
 pub fn merge_by_sizes<'a>(
     scan_tasks: BoxScanTaskIter<'a>,
+    pushdowns: &Pushdowns,
     cfg: &'a DaftExecutionConfig,
 ) -> BoxScanTaskIter<'a> {
-    Box::new(MergeByFileSize {
-        iter: scan_tasks,
-        cfg,
-        accumulator: None,
-    })
+    if let Some(limit) = pushdowns.limit {
+        // If LIMIT pushdown is present, perform a more conservative merge using the estimated size of the LIMIT
+        let mut scan_tasks = scan_tasks.peekable();
+        let first_scantask = scan_tasks
+            .peek()
+            .and_then(|x| x.as_ref().map(|x| x.clone()).ok());
+        if let Some(first_scantask) = first_scantask {
+            let estimated_bytes_for_reading_limit_rows = first_scantask
+                .as_ref()
+                .estimate_in_memory_size_bytes(Some(cfg))
+                .and_then(|est_materialized_bytes| {
+                    first_scantask
+                        .as_ref()
+                        .approx_num_rows(Some(cfg))
+                        .map(|approx_num_rows| {
+                            (est_materialized_bytes as f64) / approx_num_rows * (limit as f64)
+                        })
+                });
+            if let Some(limit_bytes) = estimated_bytes_for_reading_limit_rows {
+                return Box::new(MergeByFileSize {
+                    iter: Box::new(scan_tasks),
+                    cfg,
+                    target_upper_bound_size_bytes: (limit_bytes * 2.) as usize,
```

Comment:
i think this should be `limit_bytes * 1.5` since it should be `+ or - limit_bytes/2`

Reply:
Changed